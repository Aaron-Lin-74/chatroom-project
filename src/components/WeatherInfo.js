import React, { useState, useEffect } from 'react'

function WeatherInfo() {
  const [weatherData, setWeatherData] = useState()
  useEffect(() => {
    function getLocation() {
      if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(fetchWeather, showError)
      } else {
        console.log('Geolocation is not supported by this browser.')
      }
    }

    getLocation()
  }, [])

  // Fetch weather data from third-party API after getting the location info
  function fetchWeather(position) {
    const lat = position.coords.latitude
    const lon = position.coords.longitude
    fetch(`https://fcc-weather-api.glitch.me/api/current?lat=${lat}&lon=${lon}`)
      .then((response) => response.json())
      .then((data) => {
        setWeatherData(data)
      })
      .catch((err) => console.log(err))
  }

  function showError(error) {
    switch (error.code) {
      case error.PERMISSION_DENIED:
        console.log('User denied the request for Geolocation.')
        break
      case error.POSITION_UNAVAILABLE:
        console.log('Location information is unavailable.')
        break
      case error.TIMEOUT:
        console.log('The request to get user location timed out.')
        break
      case error.UNKNOWN_ERROR:
        console.log('An unknown error occurred.')
        break
      default:
        return
    }
  }

  if (weatherData) {
    return (
      <div className='weather-container'>
        <h2>{weatherData.name}</h2>
        <div className='weather-info'>
          <div className='weather-img-wrapper'>
            <img
              src={weatherData.weather[0].icon}
              alt={weatherData.weather[0].description}
            />
          </div>
          <span>{`${parseInt(weatherData.main.temp_min)}\u00B0C - ${parseInt(
            weatherData.main.temp_max
          )} \u00B0C`}</span>
        </div>
      </div>
    )
  } else {
    return <h1>ðŸ’¬Chat Room</h1>
  }
}

export default WeatherInfo
