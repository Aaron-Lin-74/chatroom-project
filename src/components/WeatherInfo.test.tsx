import WeatherInfo from './WeatherInfo';
import { screen, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { scryRenderedComponentsWithType } from 'react-dom/test-utils';

const server = setupServer(
  rest.get('https://fcc-weather-api.glitch.me/api/current', (req, res, ctx) => {
    return res(
      ctx.json({
        name: 'City',
        weather: [{ icon: 'url for icon', description: 'weather description' }],
        main: { temp_min: 10, temp_max: 30 },
      })
    );
  })
);

// beforeAll(() => server.listen({ onUnhandledRequest: 'bypass' }));
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

const mockNavigatorGeolocation = () => {
  const clearWatchMock = jest.fn();
  const getCurrentPositionMock = jest.fn();
  const watchPositionMock = jest.fn();

  const geolocation = {
    clearWatch: clearWatchMock,
    getCurrentPosition: getCurrentPositionMock,
    watchPosition: watchPositionMock,
  };

  Object.defineProperty(global.navigator, 'geolocation', {
    value: geolocation,
    configurable: true,
  });

  return { clearWatchMock, getCurrentPositionMock, watchPositionMock };
};

describe('Test suites for WeatherInfo component', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<WeatherInfo />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render the component with h1 by default when Geolocation is not on', () => {
    const errorObject = console.log; //store the state of the object
    console.log = jest.fn(); // mock the object
    render(<WeatherInfo />);
    expect(screen.queryByText(/Chat Room/)).toBeInTheDocument();
    console.log = errorObject; // assign it back
  });

  test('should render the weather data when Geolocation is enabled', async () => {
    const { getCurrentPositionMock } = mockNavigatorGeolocation();
    getCurrentPositionMock.mockImplementation((success, rejected) =>
      success({
        coords: {
          latitude: 51.1,
          longitude: 45.3,
        },
      })
    );
    render(<WeatherInfo />);

    expect(await screen.findByRole('img')).toBeInTheDocument();
    expect(await screen.findByText('City')).toBeInTheDocument();
    expect(await screen.findByText('10°C - 30 °C')).toBeInTheDocument();
  });
});
