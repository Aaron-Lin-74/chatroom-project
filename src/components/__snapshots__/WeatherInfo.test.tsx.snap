// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suites for WeatherInfo component should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    ðŸ’¬Chat Room
  </h1>
</DocumentFragment>
`;
